feat(Combinatorics/SimpleGraph/Bipartite): Odd Cycle Theorem (A Solution to TODO)

NickAdfor opened the pull request:
Prove bipartite iff all cycles are even

This PR formalizes the classical graph theory theorem: a graph is bipartite if and only if all its cycles have even length.

Co-authored-by: Nick Adfor <nickadfor@outlook.com>
Co-authored-by: Yue Sun <enthusiastical996@gmail.com>
Co-authored-by: Aristotle AI

---

## Main results
- `SimpleGraph.bipartite_iff_all_cycles_even` : the main equivalence theorem

## Key supporting lemmas
- `SimpleGraph.even_length_iff_same_color` : parity lemma for 2-colorings
- `SimpleGraph.even_length_iff_even_bypass_length` : bypass length parity
- `SimpleGraph.even_cycle_length_of_path` : cycle construction lemma
- `SimpleGraph.bypass_eq_nil_of_closed` : closed path properties

## Proof outline
1. For bipartite graphs: use the 2-coloring to show that in a 2-coloring, a walk has even length iff its endpoints have the same color.
2. For cycles (start = end), this forces even length.
3. For the converse: if all cycles are even, construct a 2-coloring using the `two_colorable_iff_forall_loop_even` characterization.
4. Show this coloring is valid using bypass operations and path properties.

## Technical notes
- Uses `SimpleGraph.Walk` for path/cycle representation
- Handles both `IsPath` and `IsCycle` predicates
- Uses bypass operations (`takeUntil`, `dropUntil`, `bypass`) to analyze walks
- Requires `DecidableEq V` for some lemmas
- Uses classical logic for case analysis

## Reviewer notes
- The proof uses induction on walks and case analysis on colorings
- Some lemmas (like `even_length_iff_same_color`) could be extracted to more general locations
- All theorems are in the `SimpleGraph` namespace
- The `bipartite_iff_all_cycles_even` theorem provides the complete characterization

NickAdfor committed:
Update Bipartite.lean

NickAdfor committed:
Update Bipartite.lean

NickAdfor committed:
Merge branch 'master' into feat(GraphTheory/OddCycleTheorem)

NickAdfor committed:
Update Bipartite.lean

NickAdfor committed:
Merge branch 'feat(GraphTheory/OddCycleTheorem)' of https://github.com/NickAdfor/mathlib4 into feat(GraphTheory/OddCycleTheorem)

vihdzp reviewed (changes requested):
Hi, thanks for tackling this TODO! It feels like you haven't really cleaned up the AI's output very much, there's a lot of unhelpful comments, convoluted arguments and even some non-terminal simps. Could you please review this?

NickAdfor committed:
update

vihdzp commented on Mathlib/Combinatorics/SimpleGraph/Bipartite.lean:
```diff
@@ -445,4 +446,192 @@ theorem degree_le_between_add_compl (hw : w ∈ sᶜ) :
+      _ ≤ G.dist u x + (p.dropUntil x hx).length := Nat.add_le_add_left h_dist_xv_le _
+    omega
+  refine' le_antisymm h_upper _
+  exact dist_le (p.takeUntil x hx)
```
merge?

vihdzp replied:
A refine followed by exact can generally be merged into a single exact.

vihdzp reviewed (commented):
Just a selection of suggestions.

NickAdfor committed:
Delete some AI style comments

kmill commented on Mathlib/Combinatorics/SimpleGraph/Bipartite.lean:
```diff
@@ -445,4 +446,192 @@ theorem degree_le_between_add_compl (hw : w ∈ sᶜ) :
+    · assumption
+    · exact hp_len
+  rw [← h4, ← hp_len, ← h1, SimpleGraph.Walk.length_append]
+  aesop
```
Neither `dist_eq_length_takeUntil` nor `length_dropUntil_eq_dist_sub` are used. Why are they included?

kmill commented on Mathlib/Combinatorics/SimpleGraph/Bipartite.lean:
```diff
@@ -445,4 +446,192 @@ theorem degree_le_between_add_compl (hw : w ∈ sᶜ) :
+
+lemma two_colorable_iff_forall_loop_even {α : Type*} {G : SimpleGraph α} :
+    G.Colorable 2 ↔ ∀ u, ∀ (w : G.Walk u u), Even w.length := by
+    exact SimpleGraph.two_colorable_iff_forall_loop_even
```
This is a restatement of a pre-existing theorem.

kmill commented on Mathlib/Combinatorics/SimpleGraph/Bipartite.lean:
```diff
@@ -445,4 +446,192 @@ theorem degree_le_between_add_compl (hw : w ∈ sᶜ) :
+    w.bypass = SimpleGraph.Walk.nil := by
+      have h_nil : ∀ {u : V} {p : G.Walk u u}, p.IsPath → p = SimpleGraph.Walk.nil := by
+        aesop
+      exact h_nil (SimpleGraph.Walk.bypass_isPath _)
```
Lemmas purely about `Walk.bypass` should go in `Mathlib.Combinatorics.SimpleGraph.Paths`.

NickAdfor replied:
deleted

kmill reviewed (commented):
I've made some comments stopping before `even_cycle_length_of_path`. This needs significant revision, and significant human review.

NickAdfor committed:
Update Mathlib/Combinatorics/SimpleGraph/Bipartite.lean

Co-authored-by: Violeta Hernández Palacios <vi.hdz.p@gmail.com>

NickAdfor committed:
update

Removed unused lemma and updated theorem to reflect changes.

NickAdfor committed:
update

Add a new theorem related to bipartite graphs.

NickAdfor committed:
Simplify proof of even_length_iff_same_color lemma

Refactor even_length_iff_same_color lemma to simplify proof.

NickAdfor committed:
change from Bipartite to ConcreteColorings

NickAdfor committed:
Update ConcreteColorings.lean

NickAdfor committed:
add classical

NickAdfor committed:
Update ConcreteColorings.lean

NickAdfor committed:
Update ConcreteColorings.lean

Sorry, as I cannot find a way to replace induction' and refine', I decide to ignore these problem.

NickAdfor reviewed (commented):
set_option linter.style false

vihdzp commented on Mathlib/Combinatorics/SimpleGraph/ConcreteColorings.lean:
```diff
@@ -1,7 +1,7 @@
 Copyright (c) 2023 Iván Renison. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
-Authors: Iván Renison
+Authors: Iván Renison, Sun Yue, Huang Jiayi, Nick Adfor, Aristotle
```
Is there precedent for adding Aristotle in the author field?

NickAdfor replied:
From Yuri
[#mathlib4 > Discussion thread for Aristotle API @ 💬](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/Discussion.20thread.20for.20Aristotle.20API/near/543830659)

Rida-Hamadani replied:
Yuri mentioned adding Aristotle as a PR coauthor, not as a file author. As far as I understand, the purpose of the file author field is less about attribution and more about knowing who is responsible for the code, for example in order to contact them in case something broke. Aristotle is not contactable and cannot be held responsible for its own code, so it feels pointless to add it there. To my knowledge, it also has no precedent. I think it is a good idea to discuss this on the Zulip first to get a community consensus.

vihdzp commented on Mathlib/Combinatorics/SimpleGraph/ConcreteColorings.lean:
```diff
@@ -11,6 +11,7 @@ public import Mathlib.Combinatorics.SimpleGraph.Coloring
 public import Mathlib.Combinatorics.SimpleGraph.CompleteMultipartite
 public import Mathlib.Combinatorics.SimpleGraph.Hasse
 public import Mathlib.Data.Fin.Parity
+public import Mathlib.Tactic.Cases
```
Had this really not been imported already?

vihdzp commented on Mathlib/Combinatorics/SimpleGraph/ConcreteColorings.lean:
```diff
@@ -201,4 +206,130 @@ lemma two_colorable_iff_forall_loop_even {α : Type*} {G : SimpleGraph α} :
+  classical
+  constructor
+  · intro h_bip
+    have bipartite_implies_even_cycles (h : G.IsBipartite) :
```
Is this not just your goal? Why put it in a have block?

NickAdfor replied:
Okay, deleted.

vihdzp reviewed (changes requested):
This still needs a lot of human work.

vihdzp commented on Mathlib/Combinatorics/SimpleGraph/ConcreteColorings.lean:
```diff
@@ -201,4 +206,130 @@ lemma two_colorable_iff_forall_loop_even {α : Type*} {G : SimpleGraph α} :
+    w.bypass = SimpleGraph.Walk.nil := by
+    have h_nil : ∀ {u : V} {p : G.Walk u u}, p.IsPath → p = SimpleGraph.Walk.nil := by
+      aesop
+    exact h_nil (SimpleGraph.Walk.bypass_isPath _)
```
```suggestion
@[simp]
lemma bypass_eq_nil [DecidableEq V] {G : SimpleGraph V} {u : V} (w : G.Walk u u) : w.bypass = nil :=
  (isPath_iff_eq_nil _).1 (SimpleGraph.Walk.bypass_isPath _)
```
a) This should go on the `SimpleGraph/Paths` file.
b) It might also be worth having `bypass_nil` for discoverability. 
c) `isPath_iff_eq_nil` should not take an explicit argument, which is something that perhaps should be fixed in another PR.

NickAdfor replied:
I've noticed that in your PR you changed a lot of theorem names but did not change theorem statement. So I've used your name to made new update now.

mitchell-horner commented on Mathlib/Combinatorics/SimpleGraph/Bipartite.lean:
```diff
@@ -58,10 +58,6 @@ This file proves results about bipartite simple graphs, including several double
-## TODO
-
-* Prove that `G.IsBipartite` iff `G` does not contain an odd cycle.
-  I.e., `G.IsBipartite ↔ ∀ n, (cycleGraph (2*n+1)).Free G`.
```
When I wrote this TODO, I imagined a proof in terms of `Free`.

If you are going to remove it, please at least include a proof of the statement:
```
G.IsBipartite ↔ ∀ n ≥ 2, Odd n → (cycleGraph n).Free G
``` 

Feel free to adapt the full proof I left [here](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/Formalizing.20Graph.20Theory.20-.20The.20Odd.20Cycle.20Theorem.20in.20Lean/near/565123568).

NickAdfor replied:
I should say that what I've written here is equivalent to the TODO. To show this, just write this after my PR (I've updated it in):
```lean4
theorem no_odd_cycles_implies_bipartite
(h : ∀ (v : V) (c : G.Walk v v), c.IsCycle → Even c.length) :
    G.IsBipartite := (bipartite_iff_all_cycles_even G).mpr h
```

mitchell-horner reviewed (changes requested):
Cool proof. 😁

NickAdfor committed:
Update ConcreteColorings.lean

NickAdfor committed:
Update ConcreteColorings.lean

NickAdfor committed:
Update ConcreteColorings.lean

NickAdfor committed:
Update ConcreteColorings.lean

NickAdfor committed:
Update ConcreteColorings.lean

NickAdfor committed:
Update ConcreteColorings.lean

NickAdfor committed:
name order

change the order of family name and the given name

NickAdfor committed:
Update ConcreteColorings.lean

NickAdfor committed:
更新 ConcreteColorings.lean

NickAdfor committed:
Update ConcreteColorings.lean

NickAdfor committed:
Update ConcreteColorings.lean

NickAdfor committed:
Merge branch 'master' into feat(GraphTheory/OddCycleTheorem)

NickAdfor committed:
Update Mathlib/Combinatorics/SimpleGraph/ConcreteColorings.lean

Co-authored-by: SnirBroshi <26556598+SnirBroshi@users.noreply.github.com>

NickAdfor committed:
Merge branch 'master' into feat(GraphTheory/OddCycleTheorem)

SnirBroshi commented:
> Can anyone help change the `awaiting-author`?

See [How to contribute to mathlib](https://leanprover-community.github.io/contribute/index.html) for how to remove `awaiting-author` and other useful info about contributing

SnirBroshi commented:
Also please read the [style guide](https://leanprover-community.github.io/contribute/style.html) and [this nice "clean code and formalisation best practices" guide](https://github.com/fpvandoorn/LeanCourse25/blob/43078928d4f3f8925044bcfc4068a3a9ea4505df/LeanCourse25/Lectures/Lecture21After.lean#L23), and fix the style issues (specifically I noticed the indentation is wrong in some places, but these guides are good to know in general)

NickAdfor committed:
Merge branch 'master' into feat(GraphTheory/OddCycleTheorem)

NickAdfor committed:
Update Mathlib/Combinatorics/SimpleGraph/ConcreteColorings.lean

Co-authored-by: Iván Renison <85908989+IvanRenison@users.noreply.github.com>

NickAdfor committed:
Update ConcreteColorings.lean

NickAdfor committed:
Update ConcreteColorings.lean

NickAdfor committed:
Update ConcreteColorings.lean

NickAdfor committed:
Update ConcreteColorings.lean

NickAdfor committed:
Update ConcreteColorings.lean

NickAdfor committed:
Merge branch 'master' into feat(GraphTheory/OddCycleTheorem)

vlad902 commented on Mathlib/Combinatorics/SimpleGraph/ConcreteColorings.lean:
```diff
@@ -201,4 +222,143 @@ lemma two_colorable_iff_forall_loop_even {α : Type*} {G : SimpleGraph α} :
+/-- The bypass of a walk is a path -/
+@[simp]
+lemma Walk.bypass_eq_nil
+{V : Type*} [DecidableEq V] {G : SimpleGraph V} {u : V} (w : G.Walk u u) :
```
spacing is wrong, also you redefine {V} here, try to inherit from the surrounding variable block

vlad902 reviewed (commented):
I don't think you understand your own proof based on my comment [here](https://github.com/leanprover-community/mathlib4/pull/33257/changes#r2745232536). Any part of this that is AI generated should be re-written by a human where you understand every step of it.

NickAdfor committed:
Merge branch 'master' into feat(GraphTheory/OddCycleTheorem)

NickAdfor committed:
Update ConcreteColorings.lean

NickAdfor committed:
Update ConcreteColorings.lean

NickAdfor committed:
Merge branch 'master' into feat(GraphTheory/OddCycleTheorem)

NickAdfor commented:
-awaiting-author

NickAdfor commented:
awaiting-review

YaelDillies reviewed (commented):
I understand you have been using AI extensively to write the proofs. This is okay, but you must now simplify the proofs, for example by removing the useless hypotheses, `have`s, etc...

NickAdfor committed:
Update ConcreteColorings.lean

NickAdfor committed:
Update ConcreteColorings.lean

NickAdfor committed:
Update ConcreteColorings.lean

euprunin commented:
@NickAdfor 

Are you familiar with the `merely-true` project (https://github.com/merely-true/merely-true)? It's intended as a playground for LLM-generated mathematics, with very permissive contribution guidelines.

I'm not a Mathlib reviewer or maintainer, so I don't speak for the project, but I thought I'd mention it as a potentially interesting place for work like this (independently of how this PR is evaluated here).

NickAdfor commented:
> @NickAdfor 
> 
> Are you familiar with the `merely-true` project (https://github.com/merely-true/merely-true)? It's intended as a playground for LLM-generated mathematics, with very permissive contribution guidelines.
> 
> I'm not a Mathlib reviewer or maintainer, so I don't speak for the project, but I thought I'd mention it as a potentially interesting place for work like this (independently of how this PR is evaluated here).

Very interesting! I check for a while and find its main project https://arxiv.org/pdf/2603.15929. It inspires on how to build a project by interactions between merely-one human and AIs (LLM and ATP). This Odd Cycle Theorem used to be written by @HuangJiayi0409 using ChatGPT (I guess by its Lean 3 style and prompts in comments) and I took over using Aristotle. This workflow can also be seen in https://github.com/NickAdfor/The-polynomial-method-and-restricted-sums-of-congruence-classes. But I just do odds and ends but never think of a huge project like that. Also I do think that this Odd Cycle Theorem is not serious for Mathlib (for AI style) or a paper (for lack of originality). But anyway I now know from Vasily Ilin's paper that this workflow is feasible and can be polished for a huge project (maybe a lot of automation is still needed to make it suitable for 10k lines, not just merely-one theorem).

NickAdfor commented:
Details in https://github.com/HuangJiayi0409/Graph-Formalization. It should be considered as @HuangJiayi0409's private work yet not for Mathlib. This PR might not update immediately like that repo I think.
